xdp/src/netlink.rs

```diff
@@ -572,3 +734,121 @@ pub fn netlink_get_default_gateway(family: u8) -> Result<Option<RouteEntry>, io:
 
     Ok(None)
 }
+
+// Interface information structure
+#[derive(Debug, Clone)]
+pub struct InterfaceInfo {
+    pub if_index: u32,
+    pub if_name: String,
+    pub dev_type: u16,
+}
+
+// Get all interfaces via netlink (based off of FD: fd_netdev_netlink_load_table)
+pub fn netlink_get_interfaces() -> Result<Vec<InterfaceInfo>, io::Error> {
+    let socket = NetlinkSocket::open()?;
+
+    // Create request struct
+    #[repr(C)]
+    struct Request {
+        nlh: nlmsghdr,
+        ifi: ifinfomsg,
+    }
+
+    let request = Request {
+        nlh: nlmsghdr {
+            nlmsg_type: RTM_GETLINK,
+            nlmsg_flags: (NLM_F_REQUEST | NLM_F_DUMP) as u16,
+            nlmsg_len: mem::size_of::<Request>() as u32,
+            nlmsg_seq: 1,
+            nlmsg_pid: 0,
+        },
+        ifi: ifinfomsg {
+            ifi_family: AF_INET as u8,
+            ifi_type: ARPHRD_NETROM,
+            ifi_index: 0,
+            ifi_flags: 0,
+            ifi_change: 0,
+        },
+    };
+
+    // Send request
+    let request_bytes = bytes_of(&request);
+    socket.send(request_bytes)?;
+
+    // Receive and parse messages
+    let messages = socket.recv()?;
+    let mut interfaces = Vec::new();
+
+    for msg in messages {
+        // Check for netlink errors
+        if msg.header.nlmsg_type == NLMSG_ERROR as u16 {
+            let err = msg.error.unwrap();
+            if err.error != 0 {
+                return Err(io::Error::from_raw_os_error(-err.error));
+            }
+            continue;
+        }
+
+        if msg.header.nlmsg_type != RTM_NEWLINK {
+            continue;
+        }
+
+        if let Some(if_info) = parse_ifinfomsg(&msg) {
+            interfaces.push(if_info);
+        }
+    }
+
+    Ok(interfaces)
+}
+
+// Interface info parsing
+// Just parsing the interface info for:
+// - if_index
+// - if_name
+// - dev_type
+pub fn parse_ifinfomsg(msg: &NetlinkMessage) -> Option<InterfaceInfo> {
+    let ifi = parse_into_ifinfomsg(msg)?;
+
+    // Filter interface types
+    let ifi_type = ifi.ifi_type;
+    if ifi_type != ARPHRD_ETHER && ifi_type != ARPHRD_LOOPBACK && ifi_type != ARPHRD_IPGRE {
```

Comment:
This is inverse of the check in is_valid_link_update, right? Would it make sense to reuse the function?